src/Interpreters/HashJoin.cpp

```diff
@@ -1480,9 +1494,329 @@ void setUsed(IColumn::Filter & filter [[maybe_unused]], size_t pos [[maybe_unuse
+        auto new_col = src_col->column->cloneEmpty();
+        size_t prev_left_offset = 0;
+        for (size_t i = 1; i < row_replicate_offset.size(); ++i)
+        {
+            const size_t & left_offset = row_replicate_offset[i];
+            size_t rows = left_offset - prev_left_offset;
+            if (rows)
+            {
+                new_col->insertManyFrom(*src_col->column, left_start_row + i - 1, rows);
+            }
+            prev_left_offset = left_offset;
+        }
+        executed_block.insert({std::move(new_col), src_col->type, col_name});
+    }
+    if (!executed_block)
+    {
+        WriteBufferFromOwnString buf;
+        for (const auto & col : required_cols)
+        {
+            buf << col.name << ", ";
+        }
+        throw Exception(
+            ErrorCodes::LOGICAL_ERROR,
+            "required columns: {}. but not found any in left/right table. right table: {}, left table: {}",
+            buf.str(),
+            sample_right_block.dumpNames(),
+            added_columns.left_block.dumpNames());
+    }
+    LOG_TRACE(getLogger("HashJoin"), "Additional filter execute block:\n{}", executed_block.dumpContent());
+    added_columns.additional_filter_expression->execute(executed_block);
+    LOG_TRACE(getLogger("HashJoin"), "Addition filter execute result block:\n{}", executed_block.dumpContent());
+    return executed_block.getByPosition(0).column;
+}
+
+template <bool flag_per_row>
+void addFoundRowRefAll(
+    const RowRefList & row_list,
+    std::vector<RowRef> & selected_rows,
+    IColumn::Offset & current_offset,
+    KnownRowsHolder<flag_per_row> & known_rows [[maybe_unused]])
+{
+    if constexpr (flag_per_row)
+    {
+        std::unique_ptr<std::vector<KnownRowsHolder<true>::Type>> new_known_rows_ptr;
+        for (auto it = row_list.begin(); it.ok(); ++it)
+        {
+            auto row_ref = std::make_pair(it->block, it->row_num);
+            if (!known_rows.isKnown(row_ref))
+            {
+                selected_rows.emplace_back(row_ref.first, row_ref.second);
+                ++current_offset;
+                if (!new_known_rows_ptr)
+                {
+                    new_known_rows_ptr = std::make_unique<std::vector<KnownRowsHolder<true>::Type>>();
+                }
+                new_known_rows_ptr->push_back(row_ref);
+            }
+        }
+
+        if (new_known_rows_ptr)
+            known_rows.add(std::cbegin(*new_known_rows_ptr), std::cend(*new_known_rows_ptr));
+    }
+    else
+    {
+        for (auto it = row_list.begin(); it.ok(); ++it)
+        {
+            selected_rows.emplace_back(it->block, it->row_num);
+            ++current_offset;
+        }
+    }
+}
+
+/// First to collect all matched rows refs by join keys, then filter out rows which are not true in additional filter expression.
+template <
+    typename KeyGetter,
+    typename Map,
+    bool need_replication,
+    typename AddedColumns>
+NO_INLINE size_t joinRightColumnsWithAddtitionalFilter(
```

Comment:
Can't we join blocks using regular `joinRightColumns` and filter afterwards? It's quite complex code and maintaining duplicate version can be tricky

Reply:
What if we add code handling mixedExpression into `joinRightColumns` under `if (mixed_join_expression)` (with boolean template flag and `if constexpr` if it would be some overhead). Is it possible?